### PR TITLE
[wgsl-out] Enable snapshot testing for wgsl-in image test

### PR DESCRIPTION
**Pull request by Gordon-F**
Implement all statements and expressions required by image snapshot.

### CODE REVIEW

File: src/back/wgsl/writer.rs [hunk 1]
```diff
@@ -843,32 +861,69 @@ impl<W: Write> Writer<W> {
...
-                expr,
-                kind,
-                convert: _, //TODO:
-            } => {
+            Expression::As { expr, kind, .. } => {
                 let inner = func_ctx.info[expr].ty.inner_with(&module.types);
                 let op = match *inner {
                     TypeInner::Matrix { columns, rows, .. } => {
-                        format!("mat{}x{}", vector_size_str(columns), vector_size_str(rows))
+                        format!(
```
Comment (kvark) [inline src/back/wgsl/writer.rs:909]: let's write it out to the output instead of collecting into `String` first

**Review by kvark — COMMENTED** [review]
A few fixes are needed, looks great overall!

File: src/back/wgsl/writer.rs [hunk 2]
```diff
@@ -828,13 +847,12 @@ impl<W: Write> Writer<W> {
                 // TODO: comments
                 let fun_name = match level {
                     crate::SampleLevel::Auto => "textureSample",
-                    _ => {
-                        return Err(Error::Unimplemented(format!(
-                            "expression_imagesample_level {:?}",
-                            level
-                        )));
-                    }
+                    crate::SampleLevel::Zero => "textureSampleCompare",
```
[the remark this answers is not in the document]
Comment (Gordon-F) [reply src/back/wgsl/writer.rs]: Thank you! Reworked this part. But I still not completely sure that I picked up right textures function here.

File: src/back/wgsl/writer.rs [hunk 3]
```diff
@@ -819,56 +877,136 @@ impl<W: Write> Writer<W> {
...
-                // TODO: comments
-                let fun_name = match level {
-                    crate::SampleLevel::Auto => "textureSample",
-                    _ => {
-                        return Err(Error::Unimplemented(format!(
-                            "expression_imagesample_level {:?}",
-                            level
-                        )));
+                let texture_function = if let Some(depth_ref) = depth_ref {
+                    TextureFunction::SampleCompare(depth_ref)
```
Comment (kvark) [inline src/back/wgsl/writer.rs:886]: is `TextureFunction` actually helping us here? I.e in this if/else/match we could just write the function name, and then later we could grab the expressions out of `level`

File: src/back/wgsl/writer.rs [hunk 4]
```diff
@@ -819,56 +877,136 @@ impl<W: Write> Writer<W> {
...
                 self.write_expr(module, sampler, func_ctx)?;
                 write!(self.out, ", ")?;
                 self.write_expr(module, coordinate, func_ctx)?;
+
+                if let Some(array_index) = array_index {
+                    write!(self.out, ", ")?;
+                    self.write_expr(module, array_index, func_ctx)?;
+                }
+
+                if let TextureFunction::SampleBias(expr) = texture_function {
```
Comment (kvark) [inline src/back/wgsl/writer.rs:912]: let's take this out of `level` instead: `if let crate::SampleLevel::Bias(expr) = level`.
Same for other cases

**Review by kvark — APPROVED** [review]
Great, thank you for patience!

**Pull request merged**